### DISCUSSION TITLE
The Ruby openssl extension was not compiled. Missing the OpenSSL lib?

### DISCUSSION THREAD

**Discussion by nathany** [rbenv/ruby-build#1882] [Build failures] [answered]
`rbenv install 2.0.0-p247` fails on Mac OS X 10.8.4 with the error:

```
The Ruby openssl extension was not compiled. Missing the OpenSSL lib?
```

I had no problem installing **1.9.3-p448**.

Using ruby-build 20130628 via Homebrew.

**Comment by kelan** [comment]
@nathany I had this issue too, but it's working for me now.  There were 2 things I did to try to fix it:
- Update to the latest `ruby-build`.  But, it looks like you already have that.
- Install the [rbenv ruby-build plugin](https://github.com/sstephenson/ruby-build) into `~/.rbenv/plugins/`.  Curiously,I didn't have it in that plugin dir before, but still had the `rbenv install` command available.  Maybe `rbenv` has some built-in `install` command that's older than the latest plugin, or maybe I had previously installed something somewhere else that it was picking up -- I'm not sure.  Either way, getting the latest version of the plugin in that directory seemed to help.

**Comment by nathany** [comment]
@kelan That's odd. I don't have a `~/.rbenv/plugins/` at all, as I installed/updated both rbenv/ruby-build via Homebrew.

**Comment by kelan** [comment]
@nathany That's my point.  It didn't work when I didn't have the plugin installed, but did after I installed it (as suggested on the [ruby-build plugin page](https://github.com/sstephenson/ruby-build#installing-as-an-rbenv-plugin-recommended)) by doing:

```
$ git clone https://github.com/sstephenson/ruby-build.git ~/.rbenv/plugins/ruby-build
```

So, try that and see if it helps.

**Comment by nathany** [comment]
@kelan Thanks for the tip. Since I'm not needing Ruby 2.0 for a few days, I'm going to wait-and-see if the [Installing with Homebrew](https://github.com/sstephenson/ruby-build#installing-with-homebrew-for-os-x-users) option is made to work. If not, then I'll try the method you described.

**Comment by joevandyk** [comment]
This is what I'm seeing:

```
$ git clone https://github.com/sstephenson/ruby-build.git ~/.rbenv/plugins/ruby-build
Cloning into '/Users/joe/.rbenv/plugins/ruby-build'...
remote: Counting objects: 1947, done.
remote: Compressing objects: 100% (1172/1172), done.
remote: Total 1947 (delta 783), reused 1727 (delta 592)
Receiving objects: 100% (1947/1947), 267.44 KiB, done.
Resolving deltas: 100% (783/783), done.


$ ~/.rbenv/plugins/ruby-build/bin/ruby-build 2.0.0-p195 2.0.0-p195
Downloading openssl-1.0.1e.tar.gz...
-> https://www.openssl.org/source/openssl-1.0.1e.tar.gz
Installing openssl-1.0.1e...

BUILD FAILED

Inspect or clean up the working tree at /var/folders/q7/ywcv_w057m1cmpgpjkcl69ym0000gn/T/ruby-build.20130701095203.18386
Results logged to /var/folders/q7/ywcv_w057m1cmpgpjkcl69ym0000gn/T/ruby-build.20130701095203.18386.log

Last 10 log lines:
cp: 2.0.0-p195/openssl/include/openssl/opensslconf.h: No such file or directory
chmod: 2.0.0-p195/openssl/include/openssl/opensslconf.h: No such file or directory
cp: 2.0.0-p195/openssl/include/openssl/ebcdic.h: No such file or directory
chmod: 2.0.0-p195/openssl/include/openssl/ebcdic.h: No such file or directory
cp: 2.0.0-p195/openssl/include/openssl/symhacks.h: No such file or directory
chmod: 2.0.0-p195/openssl/include/openssl/symhacks.h: No such file or directory
cp: 2.0.0-p195/openssl/include/openssl/ossl_typ.h: No such file or directory
chmod: 2.0.0-p195/openssl/include/openssl/ossl_typ.h: No such file or directory
make[1]: *** [install] Error 1
make: *** [install_sw] Error 1
```

**Comment by joevandyk** [comment]
Very weird. The above (running ruby-build directly) failed, but this works: `rbenv install 2.0.0-p195`

**Comment by kelan** [comment]
@joevandyk Yeah, I wouldn't image that rbenv plugin would affect calling `ruby-build` directly.  When I said "it worked for me", I mean using `rbenv install XXX`, which also worked for you.  So that's good.

That said, I'm not sure why `ruby-build` directly doesn't work directly.  Looks like it's an openssl dependency issue, which maybe the rbenv plugin takes care of for you?

**Comment by joevandyk** [comment]
My guess is that `rbenv install XXX` sets up the environment/paths differently (and correctly) than calling `ruby-build` directly when installing openssl.

**Comment by perfectionist** [comment]
I just updated `ruby-build` again and tried `rbenv install 2.0.0-p247` and still got error on SSL.  I have already installed several other versions of ruby without error:

Versions installed

```
$ rbenv versions
* system (set by /Users/loeffler/.ruby-version)
  1.9.3-p327
  1.9.3-p429
  1.9.3-p448
  2.0.0-p195
```

Error message:

```
$ rbenv install 2.0.0-p247
Downloading ruby-2.0.0-p247.tar.gz...
-> http://ftp.ruby-lang.org/pub/ruby/2.0/ruby-2.0.0-p247.tar.gz
Installing ruby-2.0.0-p247...

BUILD FAILED

Inspect or clean up the working tree at /var/folders/f5/7ch_20911gs9632m4nnk_0m00000gn/T/ruby-build.20130702220834.25051
Results logged to /var/folders/f5/7ch_20911gs9632m4nnk_0m00000gn/T/ruby-build.20130702220834.25051.log

Last 10 log lines:
installing default gems:      /Users/loeffler/.rbenv/versions/2.0.0-p247/lib/ruby/gems/2.0.0 (build_info, cache, doc, gems, specifications)
                          bigdecimal 1.2.0
                          io-console 0.4.2
                          json 1.7.7
                          minitest 4.3.2
                          psych 2.0.0
                          rake 0.9.6
                          rdoc 4.0.0
                          test-unit 2.0.0.0
The Ruby openssl extension was not compiled. Missing the OpenSSL lib?
```

**Comment by nathany** [comment]
@perfectionist Same experience here. Did you use homebrew to install both rbenv and ruby-build?

**Comment by perfectionist** [comment]
I did not use homebrew at all.  I had installed it once but was having issues so I uninstalled it several months ago.  That was no easy task.

I am wondering if the file `~/.rbenv/plugins/ruby-build/share/ruby-build/2.0.0-p247` should have a different version of `openssl`.

**Comment by perfectionist** [comment]
To install rbenv and ruby-install I used git.

**Comment by nathany** [comment]
Good question. p247 appears to use the same version as p195, and the later installs fine.

**Comment by danielb2** [comment]
Trying rbenv for the first time. Running into the same problem with openssl missing. 

I tried `rbenv install ruby-2.0.0-p247`  and received the `The Ruby openssl extension was not compiled. Missing the OpenSSL lib?` message.

openssl was already on the system with macports. I also installed it with homebrew. No avail.

`CONFIGURE_OPTS="--with-openssl-dir=brew --prefix openssl" rbenv install 2.0.0-p247` did not solve the problem.

However, `CFLAGS='-g -O2' RUBY_CONFIGURE_OPTS=--with-openssl-dir=`brew --prefix openssl`rbenv install 2.0.0-p247`

did solve the problem.

**Comment by nathany** [comment]
@danielb2 I got the same error with `CFLAGS='-g -O2' RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.1e rbenv install 2.0.0-p247`

**Comment by jo-sm** [comment]
Copying this from my comment [here](https://github.com/sstephenson/ruby-build/issues/197#issuecomment-20992695):

> Not sure about the Mac users, but I had this issue on my new Debian virtual machine and was perplexed as to why it was occurring when I had `openssl` installed. I researched a little bit, and eventually wound up with the solution of installing the development `ssl` library. This is what I did:
> 
> ```
> # apt-get install libssl-dev
> $ rbenv install ruby-2.0.0-p247
> Downloading ruby-2.0.0-p247.tar.gz...
> -> http://ftp.ruby-lang.org/pub/ruby/2.0/ruby-2.0.0-p247.tar.gz
> Installing ruby-2.0.0-p247...
> Installed ruby-2.0.0-p247 to /home/joshua/.rbenv/versions/2.0.0-p247
> ```
> 
> I assume you can do the same thing with `brew` but I don't know for sure. Hope this helps!

**Comment by nathany** [comment]
This did the trick (unbelievably, I hadn't installed openssl via brew, though every other version of Ruby installed fine):

``` console
brew install openssl
RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.1e rbenv install 2.0.0-p247
```

**Comment by MiguelMadero** [comment]
Explicitly specifying openssl did the trick for me. Thanks @nathany

**Comment by cfurrow** [comment]
Thanks for this. Was having the same issue, but adding the options to the command worked great. Odd.

**Comment by scottmagdalein** [comment]
Attempting to install Ruby 2.0.0-p247 with rbenv and getting the same issue. I have openssl installed via brew and I'm specifying it in the 'rbenv install' command like @nathany suggests. Still get the same error.

**Comment by cfurrow** [comment]
@scottmagdalein I haven't tried p247 yet, I was installing p195 at the time. Does that work for you?

**Comment by MiguelMadero** [comment]
it worked for me for 247

miguel

**Comment by scottmagdalein** [comment]
@cfurrow That doesn't work either. Same error.

Full error:

```
$ RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.1e rbenv install 2.0.0-p195
Downloading ruby-2.0.0-p195.tar.gz...
-> http://ftp.ruby-lang.org/pub/ruby/2.0/ruby-2.0.0-p195.tar.gz
Installing ruby-2.0.0-p195...

BUILD FAILED

Inspect or clean up the working tree at /var/folders/v1/f8v8hqk12cdcdjwp660mz8fm0000gp/T/ruby-build.20130802163339.47002
Results logged to /var/folders/v1/f8v8hqk12cdcdjwp660mz8fm0000gp/T/ruby-build.20130802163339.47002.log

Last 10 log lines:
installing default gems:      /Users/scottmagdalein1/.rbenv/versions/2.0.0-p195/lib/ruby/gems/2.0.0 (build_info, cache, doc, gems, specifications)
                          bigdecimal 1.2.0
                          io-console 0.4.2
                          json 1.7.7
                          minitest 4.3.2
                          psych 2.0.0
                          rake 0.9.6
                          rdoc 4.0.0
                          test-unit 2.0.0.0
The Ruby openssl extension was not compiled. Missing the OpenSSL lib?`
```

**Comment by mjstrasser** [comment]
I am having the same problem on OS X 10.7.5. I installed `rbenv`, `ruby-build` and `openssl` using Homebrew and removed (hopefully) all vestiges of rvm and MacPorts. I have set the environment variable `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.1e` and got the following outcomes:
- `rbenv install 1.9.3-p448` succeeds
- `rbenv install 2.0.0-p247` fails with the _Missing the OpenSSL lib?_ error message
- `rbenv install 2.0.0-p195` fails with the _Missing the OpenSSL lib?_ error message

**Comment by mjstrasser** [comment]
I have now succeeded in installing Ruby 2.0.0-p247 on OS X using `rbenv install` (i.e. `ruby-build`). 

I found references to a library under `/Users/mike/.rvm`. The reference was in the file `/usr/local/Cellar/openssl/1.0.1e/include/openssl/opensslconf.h` so I must have installed this version using Homebrew before deleting RVM (or something). So I uninstalled and reinstalled OpenSSL via Homebrew:

```
brew uninstall openssl
brew install openssl
```

After that, `rbenv install 2.0.0-p247` succeeded. There was no need to set `RUBY_CONFIGURE_OPTS` in the environment. I think `ruby-build` checks for the Homebrew version on OS X and sets the option for you.

**Comment by scottmagdalein** [comment]
Thanks for commenting here and reminding me that I need to update as well.

I got Ruby 2.0.0-p247 installed, finally, after `brew uninstall openssl` and then `brew install openssl` to clean and reinstall it. Apparently, when I ran `brew install openssl` the first time when I started the 2.0 install process with rbenv, it didn't get done right even though brew was telling me it was up to date (1.0.1e).

**Comment by runk** [comment]
Got the same issue, `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/opt/local/ rbenv install 2.0.0-p247` did a trick for me.

**Comment by troyharvey** [comment]
Thanks @nathany, this worked for me. 

```
$ brew update
$ brew install openssl
$ RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.1e rbenv install 2.0.0-p247
$ rbenv global 2.0.0-p247
```

**Comment by brownmike** [comment]
Ditto:

> Thanks @nathany, this worked for me.
> 
> ```
> $ brew update
> $ brew install openssl
> $ RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.1e rbenv install 2.0.0-p247
> $ rbenv global 2.0.0-p247
> ```

**Comment by MiguelMadero** [comment]
This might help
https://coderwall.com/p/n9bnug

Summary:

```
brew link openssl --force
```

**Comment by perfectionist** [comment]
Well, I got `2.0.0-p247` to install without home-brew. I am wondering if there is a change to the build instructions within `rbenv install` or my tweak to `RUBY_CONFIGURE_OPTS`.  Here is what I did:

First I found that the correct version of `openssl` was already installed in a previous version of ruby (2.0.0-p195) installed with `rbenv`.  To prove it I cd'ed to directory and did:

```
$ ./openssl version
OpenSSL 1.0.1e 11 Feb 2013
```

So I typed in:

```
RUBY_CONFIGURE_OPTS=--with-openssl-dir=/Users/loeffler/.rbenv/versions/2.0.0-p195/openssl
```

I double checked installed versions of ruby 

```
 $ rbenv versions
  * system (set by /Users/loeffler/.ruby-version)
   1.9.3-p327
   1.9.3-p429
   1.9.3-p448
   2.0.0-p195
```

Then did the `rbenv install`

```
$ rbenv install 2.0.0-p247
Downloading openssl-1.0.1e.tar.gz...
-> https://www.openssl.org/source/openssl-1.0.1e.tar.gz
Installing openssl-1.0.1e...
Installed openssl-1.0.1e to /Users/loeffler/.rbenv/versions/2.0.0-p247

Downloading ruby-2.0.0-p247.tar.gz...
-> http://ftp.ruby-lang.org/pub/ruby/2.0/ruby-2.0.0-p247.tar.gz
Installing ruby-2.0.0-p247...
Installed ruby-2.0.0-p247 to /Users/loeffler/.rbenv/versions/2.0.0-p247
```

I double checked and `openssl` version 1.0.1e was installed in `/Users/loeffler/.rbenv/versions/2.0.0-p247/openssl/bin` 

Bottom Line: NO brew required.

**Comment by cabhishek** [comment]
This worked for me ... 

$ brew update
$ brew install openssl
$ RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.1e rbenv install 2.0.0-p247
$ rbenv global 2.0.0-p247

**Comment by xaprb** [comment]
On my CentOS server, `yum install openssl-devel` solved the issue.

**Comment by lauradhamilton** [comment]
Had same problem. yum install openssl-devel fixed it for me too.

**Comment by O1O1O1O** [comment]
For Ubuntu you'll need:

```
sudo apt-get install libssl-dev
```

This worked a treat for me installing p247 on my virgin Ubuntu 13.04 (Raring Ringtail) system.

**Comment by agis** [comment]
I'm on Mountain Lion 10.8.5 (no rbenv installed) and it still fails:

``` bash
$ ruby-build
ruby-build 20131008
$ /usr/local/opt/openssl/bin/openssl version
OpenSSL 1.0.1e 11 Feb 2013
$ echo RUBY_CONFIGURE_OPTS
-with-openssl-dir=/usr/local/opt/openssl
$ echo CONFIGURE_OPTS
-with-openssl-dir=/usr/local/opt/openssl
$ ruby-build 2.1.0-preview1 /opt/rubies/ruby-2.1.0-preview1
```

and I get:

``` bash
Installing ruby-2.1.0-preview1...

BUILD FAILED

Inspect or clean up the working tree at /tmp/ruby-build.20131017130407.79080
Results logged to /tmp/ruby-build.20131017130407.79080.log

Last 10 log lines:
installing default gems:      /opt/rubies/ruby-2.1.0-preview1/lib/ruby/gems/2.1.0 (build_info, cache, doc, gems, specifications)
                              bigdecimal 1.2.1
                              io-console 0.4.2
                              json 1.7.7
                              minitest 4.7.5
                              psych 2.0.1
                              rake 0.9.6
                              rdoc 4.1.0.preview.1
                              test-unit 2.1.0.0
The Ruby openssl extension was not compiled. Missing the OpenSSL lib?
```

**Comment by maga** [comment]
sudo apt-get install libssl-dev
works with ubuntu 12.04 too
thanks!

**Comment by ragle** [comment]
`$ apt-get install libssl-dev`

Fixes this in Ubuntu 13.10 saucy as well. Cheers.

**Comment by nathany** [comment]
For OS X users @MiguelMadero's suggestion to use `brew link openssl` is much easier to remember than RUBY_CONFIGURE_OPTS, but I had to unlink it to install older versions of Ruby.

I finally decided to make the switch to [ruby-install](https://github.com/postmodern/ruby-install) which makes use of Homebrew. Pretty happy with that decision so far.

**Comment by mislav** [comment]
Thanks for the helpful tips, everyone. There are many people here that had the same error message, but from multiple causes. I'm sorry that many of you lost your time over this.

The main problem is that without certain dependencies on some platforms, building specific Rubies will fail with unhelpful error messages. For example, no `libssl-dev` on Ubuntu will lead to the above described problems. I've documented what I consider to be a [sane build environment](https://github.com/sstephenson/ruby-build/wiki#suggested-build-environment) on our wiki in hopes more people follow these instructions up front and avoid such pitfalls altogether.

Ideally, ruby-build would detect whether you have openssl/yaml/readline headers on your system and abort early if you're missing a dependency. However, I don't know how to do such detection reliably and cross-platform.

Now, @nathany's OS X problems are of a different nature. On a Mac, Rubies should either use system openssl that's always present, or download+compile their own if they need a newer version. Same goes for libyaml, although recent Ruby versions bundle their own yaml if system one is not found. We did add Homebrew integration to speed up builds if openssl or libyaml are available, but even without Homebrew these builds should not fail. In any case, explicitly doing this is not necessary with newer versions of ruby-build:

```
RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/...
```

However, I still don't understand the cause of the original issue. @nathany was using a recent enough ruby-build version for there to not be a problem with openssl. This leads me to believe there were two possible causes:
1. It might be that Homebrew's `openssl` was broken somehow and needed a reinstall (some people reported here that it helped)
2. It might be that you had a recent version of ruby-build installed via Homebrew but an outdated version installed in `~/.rbenv/plugins`, and `rbenv install` would use the latter.

@joevandyk Your issue when you ran ruby-build directly was caused by you trying to use a relative path as build destination and that was broken until a fix for #110 landed:

```
$ ruby-build 2.0.0-p195 2.0.0-p195
```

I'm going to close this issue since I've done a lot to document this and because I can't reproduce the original issue for some time now.

**Comment by nathany** [comment]
@mislav Thanks for the research and for updating the Wiki.

I'm not sure, but this may have something to do with the issue:

``` console
$ which openssl
/Applications/Postgres.app/Contents/MacOS/bin/openssl
```

Though it reports the same version of openssl as installed by homebrew:

``` console
openssl version
WARNING: can't open config file: /Users/mattt/Desktop/PostgresApp/Postgres/Vendor/postgres/ssl/openssl.cnf
OpenSSL 1.0.1e 11 Feb 2013
```

My .zshrc has the following statement as per the [documentation](http://postgresapp.com/documentation#toc_1):

```
PATH="/Applications/Postgres.app/Contents/MacOS/bin:$PATH"
```

I'm still using Postgres 9.2.4.1 (13). I haven't mucked about with my path and tried reinstalling ruby-build to verify if this was the cause. And if it is, I'm not sure if it still would be with PostgresApp 9.3.0.

/cc @mattt

**Comment by nathany** [comment]
- OS X 10.9 Mavericks
- brew for rbenv: stable 0.4.0, HEAD
- brew for ruby-build: stable 20131024, HEAD
- no ~/.rbenv/plugins folder (fresh install)

### With brew for openssl: stable 1.0.1e (but keg-only, not linked)

rbenv install 2.0.0-p247 # works fine now. thanks for the fix!

### Without brew install openssl, and without PostgresApp in my path:

/usr/bin/openssl # OpenSSL 0.9.8y 5 Feb 2013

rbenv install 2.0.0-p247 # Downloading openssl-1.0.1e.tar.gz... also works fine.

### Without brew install openssl, but PostgresApp 9.2.4.1 in my path:

rbenv install 2.0.0-p247 # The Ruby openssl extension was not compiled. Missing the OpenSSL lib?
rbenv install 1.9.3-p448 # still works fine

This is the situation I originally reported, and I think it confirms that [PostgresApp](http://postgresapp.com/)'s copy of openssl was somehow getting in the way.

**Comment by nathany** [comment]
This is fixed in a later release of PostgresApp https://github.com/PostgresApp/PostgresApp/releases/tag/9.2.4.3 as per issue https://github.com/PostgresApp/PostgresApp/issues/116.

So if anyone on OS X experiences this issue, either update to the latest ruby-build and brew install openssl, or update PostgresApp.

**Comment by mislav** [comment]
Wow @nathany, thanks for the investigative work. It's really insane that they suggested users edit their PATH that way. I'm glad that this is now fixed on all sides.

**Comment by jaredbeck** [comment]
I love how when a build fails, it shows the last 10 log lines.

On ubuntu 12.04, installing `libssl-dev` worked for me.

@mislav it's not cross-platform like you wanted, but you could check `dpkg -S libssl-dev` to provide an earlier warning.

**Comment by diegossilveira** [comment]
I had the same issue on CentOS 6.4.

Installing the openssl-devel package (`yum install openssl-devel`) worked for me!

`rbenv install 2.0.0-p247` succeeded!

**Comment by tristang** [comment]
Thanks @joshuasmock. Installing libssl-dev fixed the problem for me on Ubuntu 13.10.

**Comment by banderson623** [comment]
On OS X (as @nathany suggested), I had to use `brew upgrade ruby-build` to get my ruby-build up to date, this immediately installed openssl:

```
==> Upgrading 1 outdated package, with result:
ruby-build 20140110.1
==> Upgrading ruby-build
==> Installing ruby-build dependency: openssl
==> Downloading https://www.openssl.org/source/openssl-1.0.1f.tar.gz
...
```

**Comment by AlexVPopov** [comment]
I confirm that on Ubuntu 13.10, as suggested by @joshuasmock and confirmed by @tristang, the issue is solved by installing `libssl-dev`.

**Comment by namavivarjita** [comment]
I had the same issue on Debian 7.

Updating the openssl package (apt-get install openssl) worked for me!

rbenv install 2.0.0-p451 succeeded!

**Comment by soutoner** [comment]
I had the same problem on Lubuntu 13.10

It was solved with `apt-get install libssl-dev` because looking at synaptic i already had the `openssl` package.

Thank you so much!

**Comment by mylons** [comment]
For most people it's solved by installing the ssl-dev package, but for a minority who didn't solve their issue that way I give you my solution:

  I was missing libyaml and readline from:

```
yum install gcc-c++ glibc-headers openssl-devel readline libyaml-devel readline-devel zlib zlib-devel
```

which is in the wiki for ruby-build (https://github.com/sstephenson/ruby-build/wiki)

**Comment by Russell-IO** [comment]
for Mavericks, the configure opts should be:
 CONFIGURE_OPTS="--with-openssl-dir=`brew --prefix openssl`"

not
 CONFIGURE_OPTS="--with-openssl-dir=brew --prefix openssl"

without the `` around the code, the variable is set to command, with the `` the command is run then passed into the variable, which is the desired result to correctly pass the information to rbenv :)

**Comment by bucketsize** [comment]
installing libssl-dev got it to work!

**Comment by petems** [comment]
```
CONFIGURE_OPTS="--with-openssl-dir=`brew --prefix openssl`
```

 worked for me :+1:

**Comment by ghaydarov** [comment]
I had the same problem on Ubuntu 14.04 LTS.

Running sudo apt-get install openssl did not solve it. Instead running sudo apt-get install libssl-dev SOLVED it.
Thanks guys.

**Comment by thinktainer** [comment]
CONFIGURE_OPTS="--with-openssl-dir=`brew --prefix openssl` rbenv install 2.1.2

worked for me on Mavericks

**Comment by mislav** [comment]
Please note that with more recent versions of ruby-build, the explicit CONFIGURE_OPTS is not needed anymore. If you're sure that you're on latest ruby-build and openssl _didn't_ build without CONFIGURE_OPTS, please submit that as an issue.

**Comment by kentoj** [comment]
On Mac OS X Mavericks I used rbenv to install ruby 2.1.2 and it must not have pulled openssl because I get the "Missing the OpenSSL lib" message. I had the problem with installing 2.1.0 as well.

After using the command as shown by Russell-IO above I had success:
CONFIGURE_OPTS="--with-openssl-dir=`brew --prefix openssl`" rbenv install 2.1.2

I'll submit it as an issue.

**Comment by cfy-github** [comment]
On Mac OS X Mavericks 10.9.2,  I used homebrew to install the latest openssl and I used rbenv to install ruby 2.1.2,   the build FAILED with the message "The Ruby openssl extension was not compiled. Missing the OpenSSL lib?". 

After using the command:
CONFIGURE_OPTS="--with-openssl-dir=`brew --prefix openssl`" rbenv install 2.1.2
The install succeeded

So I submit it as an issue.

**Comment by bengrunfeld** [comment]
I had to use `--force` with `brew link` for it to work, but after that, everything was happy.

```
$ brew link openssl 
Warning: openssl is keg-only and must be linked with --force
Note that doing so can interfere with building software.
$  brew link openssl --force
Linking /usr/local/Cellar/openssl/1.0.1i... 1140 symlinks created
$  rbenv install 2.1.2      
Downloading ruby-2.1.2.tar.gz...
$  npm -v
1.4.23
```

**Comment by ArniDexian** [comment]
I WISH to help every person, who doesn't like any problems with package installations, and I advise you to forget about all package managers exclude MacPorts - the tools, that helps me any time with similar problems. At this time I just type in console:

> sudo port install ruby

...and all errors fall away.

**Comment by houen** [comment]
I would advise against ArniDexian's advice above.
@ArniDexian Thank you for trying to help. I would, however, advise against using MacPorts, and would advise others to not follow ArniDexian's advice. I have had nothing but trouble with MacPorts when compared to HomeBrew. If you want to, of course go ahead, but now you are warned

**Comment by ghost** [comment]
OSX 10.10.1  openssl 1.0.1j installed ( w brew) , ruby 2.1.4 installed
needed to install 2.1.5 ...
what MiguelMadero commented on Jul 26, 2013  help me...

> RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.1j rbenv install 2.1.5

**Comment by CoreyCole** [comment]
I didn't read the whole thread, but this worked to fix the same issue for me.

```
cd
git clone git://github.com/sstephenson/rbenv.git .rbenv
echo 'export PATH="$HOME/.rbenv/bin:$PATH"' >> ~/.bashrc
echo 'eval "$(rbenv init -)"' >> ~/.bashrc
exec $SHELL

git clone git://github.com/sstephenson/ruby-build.git ~/.rbenv/plugins/ruby-build
echo 'export PATH="$HOME/.rbenv/plugins/ruby-build/bin:$PATH"' >> ~/.bashrc
exec $SHELL

rbenv install 2.1.5
rbenv global 2.1.5
ruby -v
```

**Comment by Krylon360** [comment]
for CentOS 6.4; if you've installed the openssl-1.0.1e.el6_5 rpm, ftp://ftp.pbone.net/mirror/ftp.centos.org/6.5/updates/x86_64/Packages/openssl-1.0.1e-16.el6_5.4.0.1.centos.x86_64.rpm. I had to install this package due to a dependency from another project I'm working on. When Ruby; or ruby-build/install would attempt to resolve the openssl-devel by yum installing it; it would error. You will need the openssl-devel package as well.
sudo yum install -y ftp://ftp.pbone.net/mirror/ftp.centos.org/6.5/updates/x86_64/Packages/openssl-devel-1.0.1e-16.el6_5.4.0.1.centos.x86_64.rpm

Once I was able to get the correct devel package installed; @CoreyCole's directions worked just fine!

**Comment by florianb** [comment]
@mislav I'd like to suggest a check based upon `ldconfig -p | grep libssl` as it is mentioned on this [serverfault post](http://serverfault.com/questions/54736/how-to-check-if-a-library-is-installed).

If appreciated i'd like trying to create a pull-request for that issue.

**Comment by mislav** [comment]
@florianb That's a start, except it doesn't work on OS X.

Pull requests welcome of course.

**Comment by defaye** [comment]
`sudo apt-get install libssl-dev` works on ubuntu 15.04 :+1:

**Comment by mingliangfeng** [comment]
Some basic libs are required, recommend this:

```
sudo apt-get install git-core curl zlib1g-dev build-essential libssl-dev libreadline-dev libyaml-dev libsqlite3-dev sqlite3 libxml2-dev libxslt1-dev libcurl4-openssl-dev python-software-properties libffi-dev
```

**Comment by andreynering** [comment]
I had problems trying to install Ruby 2.2.2 on Fedora 22, but the following worked for me:

```
dnf install openssl-devel
rbenv install 2.2.2
rbenv global 2.2.2
```

**Comment by charneykaye** [comment]
Ubuntu 14:

```
sudo apt-get install libssl-dev
```

**Comment by emckenna** [comment]
patch worked for me when installing 2.2.2 and 2.2.3 on Arch using openssl 1.0.2.g-3

**Comment by hugobast** [comment]
I'm sorry to say this but on macOS 10.12 with latest ruby-build this is still an issue. I brew installed openssl and I still have a hell of a time in making it play nice with `rbenv install`.

It's really unfortunate but this: 

```
CONFIGURE_OPTS="--with-openssl-dir=`brew --prefix openssl`"
```

... worked ...

**Comment by PascalPixel** [comment]
macOS 10.13 High Sierra Beta 2
```CONFIGURE_OPTS="--with-openssl-dir=`brew --prefix openssl`" rbenv install 2.3.4```
Thank you @hugobast

**Comment by menjilx** [comment]
@hugobast 

Will try your solution as well, I encountered same error during installation as well.

And It works! ;) Thanks.

**Comment by dubcl** [comment]
Hi, for Debian Stretch and ruby 2.2.3 install:

apt-get install libssl1.0-dev

not libssl-dev, then apply this:

https://github.com/rbenv/ruby-build/issues/834#issuecomment-160627207

**Comment by andersonmills** [comment]
For macports, this worked for me...

```RUBY_CONFIGURE_OPTS="--with-openssl-dir=/opt/local/" rbenv install 2.4.1```

**Comment by bg-scro** [comment]
`git clone https://github.com/sstephenson/ruby-build.git ~/.rbenv/plugins/ruby-build` worked for me!

Followed that with `rbenv install 2.4.0` and everything is hunky-dory.

**Comment by zackkenyon** [comment]
when I installed rbenv through brew, I got this message
```
This formula is keg-only, which means it was not symlinked into /usr/local,
because Apple has deprecated use of OpenSSL in favor of its own TLS and crypto libraries.

If you need to have this software first in your PATH run:
  echo 'export PATH="/usr/local/opt/openssl/bin:$PATH"' >> ~/.bash_profile

For compilers to find this software you may need to set:
    LDFLAGS:  -L/usr/local/opt/openssl/lib
    CPPFLAGS: -I/usr/local/opt/openssl/include
```

have you tried adding these flags to rbenv install?

**Comment by gregorvand** [comment]
using `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/[latest version] rbenv install [version]` worked for me finally..

**Comment by wmpeng** [comment]
> using `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/[latest version] rbenv install [version]` worked for me finally..

Works for me. macOS 10.14.6

**Comment by TorvaldsDB** [comment]
> using `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/[latest version] rbenv install [version]` worked for me finally..

First, thank you so much. I failed at the  beginning because I copied the commands above guys' . You help me out of the trouble.

Here the exactly what we should do: 

```ruby
# look into your real openssl version, as for me is `1.0.2s`
ls /usr/local/Cellar/openssl/
# then to confirm the ruby version you need, as for me is `2.3.3`
# implement the command
RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.2s rbenv install 2.3.3
```
God bless you.

**Comment by garydamm** [comment]
I was having this issue with OSX 10.15.3. The version of openssl that homebrew installed was  1.1 and I was trying to install ruby 2.3.8, which did not work. 

I eventually got it to work using openssl version 1.0.2.

```RUBY_CONFIGURE_OPTS=--with-openssl-dir=/usr/local/Cellar/openssl/1.0.2 rbenv install 2.3.8```

**Comment by MilanVilov** [comment, marked as answer]
I had this issue and official rbenv documentation as well as installation of a new version of ruby (2.7.1) solved it.
https://github.com/rbenv/ruby-build/wiki#updating-ruby-build

**Comment by tom-barone** [comment]
Had the same issue trying to install ruby 3.2.2 on Ubuntu 22.04.2. Fixed it by adding:

`deb http://security.ubuntu.com/ubuntu focal-security main` to `/etc/apt/sources.list` and running

`sudo apt update && sudo apt-get install libssl1.1`
